experimental/schoology/export-assignment-data.py

```diff
@@ -0,0 +1,269 @@
+        df.to_csv(self._create_file_name(base_name), mode="w", index=False)
+
+    def _get_students(self):
+        # Pull data from Schoology
+        student_list = self.sc_client.get_students()
+
+        # Reshape for our desired output
+        def dto_to_student(student):
+            return {
+                "uid": student.uid,
+                "school_id": student.school_id,
+                "school_uid": student.school_uid,
+                "name_first": student.name_first,
+                "name_middle": student.name_middle,
+                "name_last": student.name_last,
+                "username": student.username,
+                "email_address": student.primary_email,
+            }
+
+        return [dto_to_student(s) for s in student_list]
+
+    def _get_sections(self):
+        # TODO: planning to build a fresh Schoology API package, at which time we
+        # can delete the following extension method
+        from schoolopy.models import Section
+
+        def get_sections_for_course(Self):
+            resource = "courses/{}/sections".format(Self.id)
+
+            # TODO: make sure this handles situation where there are no sections for a course
+            return [Section(raw) for raw in self.sc_client.get(resource)["section"]]
+
+        schoolopy.models.Course.get_sections = get_sections_for_course
+        del get_sections_for_course
+        # End extension
+
+        # Pull data from Schoology
+        course_list = self.sc_client.get_courses()
+
+        section_list = []
+        for c in course_list:
+            # Filter only sections for the grading periods we care about.
+            section_list += [
+                c
+                for c in c.get_sections()
+                for gp in self.grading_periods_lookup
+                if gp in c.grading_periods
+            ]
+
+        # Reshape for our desired output
+        def dto_to_section(section):
+            return {
+                "id": section.id,
+                "course_id": section.course_id,
+                "school_id": section.school_id,
+                "course_title": section.course_title,
+                "course_code": section.course_code,
+                "section_title": section.section_title,
+                "grading_period_ids": section.grading_periods,
+                "grading_period_names": [
+                    self.grading_periods_lookup[gp] for gp in section.grading_periods
+                ],
+                "enrollments": {
+                    e["id"]: e["uid"]
+                    for e in self.sc_client.get_enrollments(section.id)
+                },
+            }
+
+        return [dto_to_section(s) for s in section_list]
+
+    def _full_name(self, student):
+        names = [student["name_first"], student["name_middle"], student["name_last"]]
+        return " ".join(x for x in names if x)
+
+    def _find_student(self, students, uid):
+        return next(s for s in students if s["uid"] == uid)

```

Comment:
If you haven't seen this yet, it is like the C# `.FirstOrDefault(x => x["uid"] == uid)`. Arguably this should protect against no match found by having a try / catch on `StopIterator`. In theory it should never happen, but probably want to handle gracefully on the 1 in a million(???) chance it _does_ occur.